elyra/pipeline/component_catalog.py

```diff
@@ -195,36 +315,75 @@ class ComponentCache(SingletonConfigurable):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
-        self._cache_task_queue = Queue()
+        self._cache_queue = Queue()
+        self._manifest_queue = Queue()
+
+        manifest_dir = kwargs.get('component_manifest_dir', jupyter_core.paths.jupyter_runtime_dir())
+        self.manifest_filename = os.path.join(manifest_dir, "elyra-component-manifest.json")
 
-        # Start a thread to manage updates to the component cache
-        self._cache_manager = CacheUpdateManagerThread(self.log, self._component_cache, self._cache_task_queue)
+        # Set up watchdog for manifest
+        self.observer = Observer()
+        self.observer.schedule(ManifestFileChangeHandler(self), manifest_dir)
+
+        # Start a thread to manage updates to the component cache and manifest
+        self._cache_manager = CacheUpdateManagerThread(
+            self.log,
+            self._component_cache,
+            self._cache_queue,
+            self._manifest_queue,
+            self.manifest_filename
+        )
         self._cache_manager.name = "CacheUpdateManager"
         self._cache_manager.start()
 
-        self._build_cache()
-
-    def _build_cache(self):
+    def load(self):
+        """
+        Completes a series of actions during system startup, such as creating
+        the component manifest file and triggering the build of the component
+        cache for existing ComponentCatalog metadata instances.
         """
-        Reads through all ComponentCatalogMetadata instances to build the initial component cache
+        # Proceed only if singleton instance has been created
+        if self.initialized:
+            # Create/overwrite manifest file and start its watchdog
+            self.set_up_manifest_file()
+            self.observer.start()
+
+            # Fetch all component catalog instances and trigger their add to
+            # the component manifest, and consequently, the component cache
+            catalogs = MetadataManager(schemaspace=ComponentCatalogs.COMPONENT_CATALOGS_SCHEMASPACE_ID).get_all()
+            for catalog in catalogs:
+                self.update_manifest(catalog, 'modify')
+
+    def set_up_manifest_file(self):
         """
+        Build an empty manifest and overwrite or create the file. After this point,
+        only the CacheUpdateManagerThread is able to touch this file.
+        """
+        empty_manifest = {
```

Comment:
This seems like a good area to describe the manifest. Let's add as many type hints into this as we can.